Admin area codemirror JSON editor does not display arrays

tazo90 opened the pull request:
It allows to display arrays in codemirror editor.

lsbardel commented:
@tazo90 can you update this pull request so that @reupen can review?

tazo90 commented:
@lsbardel I've updated PR.

reupen commented:
Hi @tazo90,

This problem isn't too important at the moment, but comments are:

* A JSON field can be a string, number, boolean, array or 'object'. There is an existing check [here](https://github.com/quantmind/lux/pull/171/files#diff-c53c6a7f4f24f5092ab9d33efb772e59L95) similar to the new one. Instead of adding another check for arrays we should modify the existing one to stringify any JSON field value, not just objects. We probably need to modify [this line in forms.js](https://github.com/quantmind/lux/blob/d076f57158d6e43e2e57eac5bf0b4cd6ccbadbe1/luxjs/forms/form.js#L162) to store all the field info so we know whether it is a JSON field or not:
* The other problem is that the back end is appending '[]' to the field name if the value of the JSON field is an array. I don't think it should be doing this as otherwise the name of the field changes depending on the value!

Thanks

tazo90 commented:
@reupen thanks for your suggestions, it makes sense. I will do changes.

lsbardel commented:
@reupen right, we don't want that. I was not aware of that bug.

lsbardel commented:
Can this be tested?

tazo90 commented:
@lsbardel I will add tests for this obviously.

tazo90 commented:
Fixed.

This removes the braces from json arrays fields because of following issue [https://github.com/quantmind/lux/issues/197](https://github.com/quantmind/lux/issues/197).

- [x] fixed codemirror issue for json arrays 
- [x] tests

**Preview**
![json-arrays](https://cloud.githubusercontent.com/assets/3165572/11475053/a7180c58-977b-11e5-8f6d-a1d3ba41a2ce.png)

lsbardel commented on luxjs/utils.js:
```diff
@@ -245,6 +245,20 @@
+    },
+
+    /**
+     * Obtain an json object from string (if available) otherwise null
```
Obtain a JSON object from a string (if available) otherwise null

lsbardel commented on luxjs/tests/lux.js:
```diff
@@ -86,4 +86,12 @@
+        it("Check getJsonOrNone", function() {
+            var res;
+            res = lux.getJsonOrNone("{\"id\": 1}");
+            expect(res.id).toBeDefined();
```
``.toBe(1)``

tazo90 commented:
Fixed.

reupen commented on luxjs/services/luxweb.js:
```diff
@@ -92,19 +92,28 @@
+                            var jsonArrayKey = key.split('[]')[0];
+
+                            // Stringify json only if has json mode enabled
+                            if (modelType[jsonArrayKey] === 'json' && (isObject(value) || isArray(value))) {
```
Strings must be JSON-encoded too (for a JSON field).

tazo90 replied:
I added a condition for a Strings too.
